Feature/ccmp

TonyJR opened the pull request:
Support noto-emoji
## Description
add a new contextChecker for ccmp table in tokenizeText()
add a new substitutionFormat for 'lookup sub_5'

## Motivation and Context
This PR resoved multi-character emoji #511.
noto-emoji uses the 'ccmp' table to solve the 'multi-character for one emoji' problem, which is a 'Language forms' feature that should be executed before all other features. When parsing the noto-emoji font, I found that 'lookup sub_5' could not be parsed correctly, so I added a substitutionFormat to handle this issue.

## How Has This Been Tested?
The unit tests passed and I added a test case for this feature. 


## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I did `npm run test` and all tests passed green (including code styling checks).
- [x] I have added tests to cover my changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the **README** accordingly.
- [x] I have read the **CONTRIBUTING** document.

Connum commented:
This looks like a great addition! Please check why the build is failing: https://github.com/opentypejs/opentype.js/actions/runs/8388243929/job/22972963564?pr=688
This PR also implements some additional substitution formats/types, which might conflict with some other outstanding PRs, but we'll have to check.

Connum commented on src/features/applySubstitution.js:
```diff
@@ -1,3 +1,11 @@
+/*
```
Can you please remove these file headers?

TonyJR replied:
oh, of course

Connum commented on src/features/ccmp/ccmpReplacementLigatures.js:
```diff
@@ -0,0 +1,56 @@
+/*
```
Can you please remove these file headers?

Connum commented on src/features/ccmp/contextCheck/ccmpReplacement.js:
```diff
@@ -0,0 +1,22 @@
+/*
```
Can you please remove these file headers?

Connum commented on test/emoji.js:
```diff
@@ -0,0 +1,47 @@
+/*
```
Can you please remove these file headers?

TonyJR commented:
okay, i removed the file header and fixed a code error.

Connum commented:
There are still some lining issues, can you please resolve them? Then you can run `npm run lint` locally before committing in order to check if everything is OK.

Connum commented on src/features/featureQuery.js:
```diff
@@ -169,6 +169,13 @@ function chainingSubstitutionFormat3(contextParams, subtable) {
+                        const glyphIndex = contextParams.get(n);
+                        const substitution = lookup(glyphIndex);
+                        if (substitution)
+                            substitutions.concat(substitution);
```
`.concat()` returns a new array, which is not stored anywhere here? So this should cause problems with specific substitutions.
We also want to avoid `.concat()` for performance reasons (see #690).

TonyJR replied:
Thank you for your reminder. I readed #690, you are right. I'll fix this.

Connum replied:
let me resolve this conversation once I did a full review

Connum commented:
Please also add tests for the ligature substitution format 3 and the substitution type 21 (which should catch the mistake I mentioned in the comment I just added to the associated code)

TonyJR commented:
> Please also add tests for the ligature substitution format 3 and the substitution type 21 (which should catch the mistake I mentioned in the comment I just added to the associated code)

I'm trying to add tests for ligature with sub_5, it doesn't work. Maybe I need a few days to fix it. This may modify the code related to "liga" and "rlig".
Any suggestions?

TonyJR commented:
> Please also add tests for the ligature substitution format 3 and the substitution type 21 (which should catch the mistake I mentioned in the comment I just added to the associated code)

I removed substitution type 21 and ligatureSubstitutionFormat3. As an alternative, use chainSubstitutionFormat3 to process the sub5 table. Sub5 may exist in liga and ccmp, both of which are processed using ligatureSubstitutionFormat3. I added test code to ensure they work properly.

Connum commented:
Thanks for your work! Two more things, then we'll hopefully be ready for review:
1. please add the information for the newly added test fonts to the `test/fonts/LICENSE` file (alphabetic order)
2. Instead of adding a new `liga.js` test file, please add the tests for the two new lookup formats to `test/featureQuery.js` in analog form as the existing tests.

Big thanks again!

TonyJR commented:
![image](https://github.com/opentypejs/opentype.js/assets/7473679/64da383f-c7b6-4f04-b83b-cc9451ccdd07)
@Connum Do you have a todo list of the LookupTypes? I want to implement this part of the code, but I am concerned that it may overlap with your work. So, do you have any suggestions?

Connum commented:
Great to see someone being interested in contributing! There is an issue with an overview of TODOs: https://github.com/opentypejs/opentype.js/issues/374
It might not be up to date, so some things might have been implemented meanwhile. You could also search the code base for instances of "not supported" or "not yet supported".

TonyJR commented:
> Great to see someone being interested in contributing! There is an issue with an overview of TODOs: #374 It might not be up to date, so some things might have been implemented meanwhile. You could also search the code base for instances of "not supported" or "not yet supported".

Thanks for your suggestion. I'll finsh the format 5.1,5.2,5.3 in this PR, please merge them later.

TonyJR commented:
> Great to see someone being interested in contributing! There is an issue with an overview of TODOs: #374 It might not be up to date, so some things might have been implemented meanwhile. You could also search the code base for instances of "not supported" or "not yet supported".

Thanks for your suggestion. I'll finsh the format 5.1,5.2,5.3 in this PR, please merge them later.

Connum commented on src/features/ccmp/ccmpReplacementLigatures.js:
```diff
@@ -0,0 +1,47 @@
+}
+
+/**
+ * Apply Arabic required ligatures to a context range
```
Should be "Apply ccmp replacement ligatures to a context range"

Connum commented on src/features/applySubstitution.js:
```diff
@@ -30,6 +30,15 @@ function chainingSubstitutionFormat3(action, tokens, index) {
         const token = tokens[index + i];
+        if (Array.isArray(subst)) {
+            if (subst.length){
+                // TODO: replace one glyph with multiple glyphs
```
Is that something that should be in this PR?

Connum commented on src/features/applySubstitution.js:
```diff
@@ -50,14 +59,17 @@ function ligatureSubstitutionFormat1(action, tokens, index) {
     }
 }
 
+
```
we don't need this additional line break

Connum commented on src/features/ccmp/ccmpReplacementLigatures.js:
```diff
@@ -0,0 +1,47 @@
+ * @param {ContextRange} range a range of tokens
+ */
+function ccmpReplacementLigatures(range) {
+    const script = 'delf';
```
I guess this should be 'DFLT'?

Connum commented on src/features/featureQuery.js:
```diff
@@ -435,6 +532,26 @@ FeatureQuery.prototype.lookupFeature = function (query) {
                     break;
+                case '51':
+                    substitution = lookup(contextParams);
+                    if (Array.isArray(substitution) && substitution.length) {
```
Are there cases where substitution is not an array or might be empty? Or should that theoretically never happen in a well-formed font and this is just to catch any malformed ones?

Connum commented on src/features/featureQuery.js:
```diff
@@ -435,6 +532,26 @@ FeatureQuery.prototype.lookupFeature = function (query) {
+                    break;
+                case '53':
+                    substitution = lookup(contextParams);
+                    if (Array.isArray(substitution) && substitution.length) {
```
Same as the previous comment above

Connum commented on src/features/featureQuery.js:
```diff
@@ -435,6 +532,26 @@ FeatureQuery.prototype.lookupFeature = function (query) {
                         }));
                     }
                     break;
+                case '51':
```
can we get rid of the code duplication between '51' and '53'?

```
                case '51':
                case '53':
                    substitution = lookup(contextParams);
                    if (Array.isArray(substitution) && substitution.length) {
                        substitutions.splice(currentIndex, 1, new SubstitutionAction({
                            id: parseInt(substType),
                            tag: query.tag,
                            substitution,
                        }));
                    }
                    break;
```

Connum commented on test/emoji.js:
```diff
@@ -0,0 +1,39 @@
+import assert from 'assert';
```
Instead of adding a new `emoji.js` test file, please add these tests to `test/bidi.js`.

Connum reviewed (changes requested):
I added a few comments for things that need clarification or fixing.

Connum commented:
Another thing I noticed: The 'ccmp' tag was already present in the `requiredThaiFeatures`:
https://github.com/opentypejs/opentype.js/blob/6408975ed2c6e91ac8376d8be6290b2f705e1724/test/bidi.js#L114

Should this be present in `requiredArabicFeatures` and `latinFeatures` as well?
https://github.com/opentypejs/opentype.js/blob/6408975ed2c6e91ac8376d8be6290b2f705e1724/test/bidi.js#L25
https://github.com/opentypejs/opentype.js/blob/6408975ed2c6e91ac8376d8be6290b2f705e1724/test/bidi.js#L40

Or is it maybe not needed at all, because the `'ccmp'` is always applied? And if so, does the code make sure of that?

This might also relate to your comment https://github.com/opentypejs/opentype.js/pull/688#issuecomment-2031453250 ? I think you forgot to describe the issue with the language tag for us. Could it be related to the wrong `'delf'` script tag I hinted at in the review comment?
